Core/clim-basic/text-formatting.lisp

```
@@ -0,0 +1,130 @@
+        (stream-page-region stream)
+      (declare (ignore max-x max-y))
+      (values min-x min-y))))
+
+(defgeneric page-cursor-final-position (stream)
+  (:documentation "Returns two values: x and y final position for a cursor on page.")
+  (:method ((stream standard-page-layout))
+    (with-bounding-rectangle* (min-x min-y max-x max-y)
+        (stream-page-region stream)
+      (declare (ignore min-x min-y))
+      (values max-x max-y))))
+
+(defmethod slot-unbound (class (stream standard-page-layout) (slot (eql '%page-region))
+                         &aux (sheet-region (sheet-region (or (pane-viewport stream)
+                                                              stream))))
+  (when (eql sheet-region +everywhere+)
+    (let ((x2 (* 80 (text-style-width  (stream-text-style stream) stream)))
+          (y2 (* 43 (text-style-height (stream-text-style stream) stream))))
+      (setf sheet-region (make-rectangle* 0 0 x2 y2))))
+  (with-bounding-rectangle* (x1 y1 x2 y2) sheet-region
+    (macrolet ((thunk (margin edge sign direction)
+                 `(if (eql (first ,margin) :absolute)
+                      (parse-space stream (second ,margin) ,direction)
+                      (,sign ,edge (parse-space stream (second ,margin) ,direction)))))
+      (destructuring-bind (&key left top right bottom) (stream-text-margins stream)
+        (setf (%page-region stream)
+              (make-rectangle* (thunk left   x1 + :horizontal)
+                               (thunk top    y1 + :vertical)
+                               (thunk right  x2 - :horizontal)
+                               (thunk bottom y2 - :vertical)))))))
+
+(defmethod (setf sheet-region) :before (sheet-region (stream standard-page-layout))
+  (unless (region-equal sheet-region (sheet-region stream))
+    (slot-makunbound stream '%page-region)))
+
+(defun valid-margin-spec-p (margins)
+  (ignore-errors ; destructuring-bind may error; that yields invalid spec
```

I'm not a huge fan of `IGNORE-ERRORS`. It hides legitimate errors that should crash the application rather than return nil. Couldn't this actively check for the error cases and explicitly return nil in those cases?

---

what is an error case for destructuring-bind? `error`? how that would be different from ignore-errors? also this calls nothing outside this scope, so if anything causes the error it is malformed margin specification structure, hence should return nil. I'm against using ignore-errors for blocks which call anything arbitrary too.

---

Good point. I guess the best I can come up with is using something like lambda-fiddle (available on QL). However, I'm not sure that is a complete solution anyway.